internal/xdscache/v3/endpointslicetranslator.go

```diff
@@ -0,0 +1,458 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	envoy_endpoint_v3 "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	resource "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	"github.com/projectcontour/contour/internal/contour"
+	"github.com/projectcontour/contour/internal/dag"
+	envoy_v3 "github.com/projectcontour/contour/internal/envoy/v3"
+	"github.com/projectcontour/contour/internal/k8s"
+	"github.com/projectcontour/contour/internal/protobuf"
+	"github.com/projectcontour/contour/internal/sorter"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+	v1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+// RecalculateEndpointsSlice generates a slice of LoadBalancingEndpoint
+// resources by matching the given service port to the given discoveryv1.EndpointSlice.
+// endpointSliceMap may be nil, in which case, the result is also nil.
+func RecalculateEndpointsSlice(port, healthPort v1.ServicePort, endpointSliceMap map[string]*discoveryv1.EndpointSlice) []*LoadBalancingEndpoint {
```

Comment:
I'd say make this a method on EndpointSliceCache so we can name it just `RecalculateEndpoints` and it doesn't conflict with the existing one, since what it's recalculating is just (LoadBalancing) Endpoints.